File: TaeBeomShin/6. 다이나믹_프로그래밍/못생긴_수.java
import java.util.Scanner;

import static java.lang.Math.min;

public class 못생긴_수 {
    public static void main(String[] args) {
        Scanner sc= new Scanner(System.in);
        int n=sc.nextInt();

        int[] numbers=new int[n+1];

        int idx2=0,idx3=0,idx5=0;
        int next2=2,next3=3,next5=5;

        numbers[0]=1;
        for(int i=1;i<=n;i++){
            numbers[i]=min(min(next2,next3),next5);
            if(numbers[i]==next2){
                idx2++;
                next2=numbers[idx2]*2;
            }
            if(numbers[i]==next3){
                idx3++;
                next3=numbers[idx3]*3;
            }
            if(numbers[i]==next5) {
                idx5++;
                next5 = numbers[idx5] * 5;
            }
        }
        System.out.println(numbers[n-1]);
    }
}

File: yunva17/다이나믹 프로그래밍/못생긴_수.py
n = int(input())

num = [0]*n # 못생긴 수를 담기 위한 테이블
num[0]=1

i2 = i3 = i5 = 0
next2, next3, next5 = 2,3,5

for l in range(1,n):
    num[l] = min(next2, next3, next5)
    if num[l]== next2:
        i2+=1
        next2 = num[i2]*2
    if num[l]== next3:
        i3+=1
        next3 = num[i3]*3
    if num[l]== next5:
        i5+=1
        next5 = num[i5]*5

print(num[n-1])
